mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,224 @@
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "adjoint"
+
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+template <class T> T isInstanceOf(Operation &op)
+{
+    if (op.getName().getStringRef() == T::getOperationName())
+        return cast<T>(op);
+    else
+        return nullptr;
+}
+
+/// Copy the region of the adjoint operation `op` to the POI specified by the `rewriter`. Build and
+/// return the value mapping `bvm`.
+Value copyAdjointVerbatim(AdjointOp op, PatternRewriter &rewriter, IRMapping &bvm)
+{
+    Block &b = op.getRegion().front();
+    for (auto i = b.begin(); i != b.end(); i++) {
+        if (YieldOp yield = isInstanceOf<YieldOp>(*i)) {
+            assert(++i == b.end() &&
+                   "quantum.yield must be the last operation of an adjoint block");
+            return bvm.lookupOrDefault(yield->getOperand(0));
+        }
+        else {
+            rewriter.insert(i->clone(bvm));
+        }
+    }
+    assert(false && "quantum.yield must present in the adjoint region");
+}
+
+struct AdjointSingleOpRewritePattern : public mlir::OpRewritePattern<AdjointOp> {
+    using mlir::OpRewritePattern<AdjointOp>::OpRewritePattern;
+
+    /// In essence, we build a map from values mentiond in the source data flow to the values of the
+    /// program where quantum control flow is reversed. Most of the time, there is a 1-to-1
+    /// correspondence with a notable exception caused by `insert`/`extract` API asymetry.
+    mlir::LogicalResult matchAndRewrite(AdjointOp adjoint,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        LLVM_DEBUG(dbgs() << "Adjointing the following:\n" << adjoint << "\n");
+        Location loc = adjoint.getLoc();
+        MLIRContext *ctx = adjoint.getContext();
+        assert(adjoint.getRegion().hasOneBlock());
+
+        // First, copy the classical computations directly to the target POI and build the classical
+        // value mapping dictionary.
+        auto classicalMapping = ({
+            IRMapping out;
+            Block &b = adjoint.getRegion().front();
+            for (auto i = b.begin(); i != b.end(); i++) {
+                if (i->getName().getStringRef().find("quantum") != std::string::npos) {
```

Comment:
You can also find out whether an operation is a quantum operation using the following:

```
     Dialect *dialect = operation->getDialect();
     bool isUserInQuantumDialect = isa<QuantumDialect>(dialect);
```

Reply:
There are also other ways to walk instructions in a block. Does this consider nested operations?

Reply:
As another note, I had not seen this syntax before. `auto var = ({ stat; stat; expr; });` very interesting!

Reply:
> Does this consider nested operations?

No it does not, but I believe this is what do we want here. Nested operations require special care, you can check how do I deal with nested adjoint blocks for example.

> auto var = ({ stat; stat; expr; });
Expression statements or something like this. I believe this is the best (but never perfect) way to overcome some C++ weak points.

Reply:
> isa<QuantumDialect>(dialect);

Done

Reply:
> No it does not, but I believe this is what do we want here. Nested operations require special care, you can check how do I deal with nested adjoint blocks for example.

Question (which maybe @dime10 can also contribute), in MLIR what is a good way to limit the transformation in such a way that it is only applied if all operations are known. Is this transformation limited to cases where there would be no operations with nested regions except those which are explicitly handled in the if statements below?

Reply:
I suppose it would be nice to abort the pass if cannot guarantee that it is correct. But for the purely classical code we don't care if it has nested regions, we only care about the final values that go into the gates, so as long as those are produced the pass should work correctly. Note that `clone` does clone the operation with its nested regions.

It would be good to add a test case for this however, for example a classical loop whose result gets passed to a gate :)